### PR TITLE
Initial commit

### PR DESCRIPTION
**Pull request by lieut-data**
The sample plugin exercises all hooks and webapp components, and should double as a useful testbed for QA to verify plugin functionality from release to release.

Lots of documentation inline, so I'll leave that for review.

This PR was originally scheduled for https://github.com/mattermost/mattermost-plugin-sample, but after discussion, we agreed to reduce the scope of the sample plugin to just a trivial example vs. the full-featured example in this demo repository.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Demo Plugin
+
+This plugin demonstrates the capabilities of a Mattermost plugin. It includes the same build scripts as [mattermost-plugin-sample](https://github.com/mattermost/mattermost-plugin-sample), but implements all supported server-side hooks and registers a component for each supported client-side integration point. See [server/README.md](server/README.md) and [webapp/README.md](webapp/README.md) for more details. The plugin also doubles as a testbed for verifying plugin functionality during release testing.
+
+Feel free to base your own plugin off this repository, removing or modifying components as needed. If you're already familiar with what plugins can do, consider starting from [mattermost-plugin-sample](https://github.com/mattermost/mattermost-plugin-sample) instead, which includes the same build framework but omits the demo implementations.
```
Comment (jwilander) [inline README.md]: Might be good to add a short disclaimer on a new line letting people know this will only work for MM 5.2+

**Comment by jasonblais** [comment]
@lieut-data Once merged, please help reference the demo and sample plugins at developers.mattermost.com + add to our plugins repo: https://github.com/mattermost/mattermost-plugins

File: server/http_hooks.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	var response = struct {
+		Enabled bool `json:"enabled"`
+	}{
+		Enabled: !p.disabled,
+	}
+
+	responseJSON, _ := json.Marshal(response)
+
+	w.Write(responseJSON)
+}
```
Comment (jwilander) [inline server/http_hooks.go]: Might be a good idea to add a brief routing example here, I've had community/customers ask me how to do that fairly often

Comment (lieut-data) [reply server/http_hooks.go]: Good catch! This also exposed a bug with plugins handling `/subpath`: https://github.com/mattermost/mattermost-server/pull/9161.

**Pull request merged**